### PR TITLE
docker-compose local dev env with live reload

### PR DESCRIPTION
**Pull request by xnutsive**
Preface: @hamelsmu, this turned out to be quite an opinionated PR. I didn't mean to send `_dev_tools` to wreckage, rather tried to simplify the flow for the users as much as possible and came up with this.

## Adding live reload / watch to local env

We've briefly talked about this — it'd be great if users could preview their posts locally, and ideally if the posts would be automatically regenerated when the user edits them in a notebook, or even a word document. This PR implements that behavior. 

### Usage
Assuming you're in the repository root: 

```bash
docker-compose build
docker-compose up
```

### Overview of what's changed
One major concern was to make sure that the local jekyll and python environment are as close to what fastpages runs in the github action as possible. 

I've reused `_dev_tools` docker-compose files and united them into a single file with multiple services. The side effect of this is the commands are much easier to run now (no -f convert-docker-compose), and therefore `Makefile` didn't make much sense. **I've dropped _dev_tools directory.**

The `docker-compose.yml` now has these services:
- `jekyll` that runs `jekyll serve`
- `converter` that converts _word and _notebook posts once, via `_action_files/action_entrypoint.sh`
- `watcher` that watches for changes in notebooks and word docs and converts them when they change.
- `jupyter` that spins up a Jupyter notebook in the blog directory. 

Running `docker-compose up` runs all the services. You can run them individually (docker-compose up jekyll) or in any combination. 

Running `docker-compose run jekyll /bin/bash` will give you a shell into the Jekyll server.   

_Note: I've updated `CONTRIBUTING.md` and `README.md` with some notes from the readme file in `_dev_tools`_. 

I've tested this locally, and it'd be great if you could test this once over to make sure I didn't miss anything important. How does it look?

### CODE REVIEW

**Comment by hamelsmu** [comment]
This will take me some time to review properly will do this middle/late next week

**Comment by hamelsmu** [comment]
Some early feedback.  **I really, really like where this is going BTW your improvements are going to make it allot easier for everyone**.   I love the hot reloading, and I love that there is just one set of services so I only need to keep one tab open

There is some functionality I miss in the makefile that I really liked

1.  Ability to remove the container completely from the system (otherwise you are left with orphan containers)

```
docker stop fastpages-dev || true; docker rm fastpages-dev || true;
docker stop jekyll || true; docker rm jekyll || true;
```

2. Force a build **without cache** 

If I change the contents of `_action_files` I found that I need to orchestrate a build without cache, which is why I have

```
docker stop fastpages-dev || true; docker rm fastpages-dev || true;
docker-compose -f convert-docker-compose.yaml  build --no-cache
```

3.  Ability to SSH into the container

For various reasons I want to quickly SSH into the container, sometimes I want to install something there where I need access to the terminal inside the docker container.  

That is why there were these commands

```
docker-compose exec jekyll bash
docker-compose -f convert-docker-compose.yaml run --name fastpages-dev fastpages /bin/bash
```

5. Calling `make` without any commands gave me a really quick inventory of all the options I had available and was very convenient. 

# This was not in the makefile but I wish it was:

4.   **rebuild with cache**, sometimes for the same reasons I want to force a rebuild of the container because I change the notebook template fore example so I want to rebuild with cache.

5.  Something simple for restarting the server quickly, I suppose that is just means `docker-compose down` followed by `docker-compose up`

I need to do this because If you make a change to `_config.yaml` or some other items, you have to re-start the server.  Jekyll hot-reloading only works for _includes.

@xnutsive Perhaps we bring back the makefile, or if there is another way we can facilitate this for users?  

### Other Feedback (some of this I neglected myself)

- Perhaps set the Jupyter server to run without any password or token, so we can just navigate to localhost:8888 without scrolling through the logs in the terminal to find the token or the magic link

- You may have not resolved all of the links in the repo that link to `_dev_tools` since that folder has been killed in this PR (However maybe you want to bring it back now, not sure).  

- Perhaps put a link to the how to install Docker from the docs

- Make it clear that you must be in the root of the repo to run these various commands, etc.

🙇  Thank you so much

File: docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    tty: true
+    volumes:
+      - .:/data/
+
+  converter:
+    <<: *fastpages
+    command: /fastpages/action_entrypoint.sh
+
+  watcher:
+    <<: *fastpages
```
Comment (hamelsmu) [inline docker-compose.yml]: I didn't know about this sytnax before.  This is so cool!  And useful!

**Comment by xnutsive** [comment]
@hamelsmu thank you very much for reviewing this! I've rewritten the section in the README based of your questions, and I think it became much more detailed and easy to use now. 

Here are a few examples of how the things you need work with this setup: 

> Ability to remove the container completely from the system (otherwise you are left with orphan containers)
```
docker-compose rm
```

> Force a build without cache
```
docker-compose build --no-cache --force-rm
```

> Ability to SSH into the container
```
docker-compose exec jekyll /bin/bash
docker-compose exec watcher /bin/bash
```

> Calling make without any commands gave me a really quick inventory of all the options I had available and was very convenient.

The closest thing to this that the setup has is `docker-compose config`, that shows that services are there. I think the readme now covers everything the users will need, and more docker-savvy folks already know their way around docker-compose. 

> rebuild with cache, sometimes for the same reasons I want to force a rebuild of the container because I change the notebook template fore example so I want to rebuild with cache.

```
docker-compose build
```

> Something simple for restarting the server quickly, I suppose that is just means docker-compose down followed by docker-compose up
That's correct, just `docker-compose restart` if the containers were running with `docker-compose up -d` (in the background). Or just `Ctrl+C` to kill the existing containers and `docker-compose up` again. 

> Perhaps set the Jupyter server to run without any password or token, so we can just navigate to localhost:8888 without scrolling through the logs in the terminal to find the token or the magic link

I haven't done this yet, will take a look now. 

> You may have not resolved all of the links in the repo that link to _dev_tools since that folder has been killed in this PR.
You're right! 🤦🏼‍♂️ Good catch, I've grepped the repo for `_dev_tools` and fixed all the references.

> Perhaps put a link to the how to install Docker from the docs
Done! Put the link in the section of README.

> Make it clear that you must be in the root of the repo to run these various commands, etc
Also done.

**Comment by xnutsive** [comment]
I've looked up if we can launch Jupyter without the token. There's a bunch of [discussions like this](https://github.com/jupyter/notebook/issues/2254) on the topic.

TLDR: we can. If I add `--NotebookApp.token=''` flag to the jupyter command, it'll run without a password or a token by default. However, it's not a recommended practice. 

I'd say let's keep the token, and if too many users request a way to launch without the token verification, we setup a default password and note it in the readme. 

How does that sound?

**Comment by xnutsive** [comment]
The only thing I need help with is to update the TOC in README.md. It seems like it's generated automatically, I don't want to mess it up, but I've added a bunch of headings and it would make sense to add them to TOC as well.

File: README.md [hunk 2]
```diff
@@ -153,6 +154,95 @@ On social media sites like Twitter, an image preview can be automatically shown
...
+# run all services in the background
+docker-compose up -d 
+
+# stop the services
+docker-compose down
+```
+
+If you need to restart just the Jekyll server, and it's running in the background — you can do `docker-compose restart jekyll`.
+
+_Note that the blog won't autoreload on change, you'll have to hit refresh manually._
```
Comment (hamelsmu) [inline README.md]:
```suggestion
_Note that the blog won't autoreload on change, you'll have to refresh your browser manually._
```
Only b/c I was confused what kind refresh was being referred to
